server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -172,7 +203,10 @@ protected SearchHit nextDoc(int doc) throws IOException {
                     ctx,
                     leafSourceLoader,
                     leafIdLoader,
-                    rankDocs == null ? null : rankDocs.get(doc)
+                    rankDocs == null ? null : rankDocs.get(doc),
+                    circuitBreaker,
+                    enoughBytesOrLastDocInLeaf,
+                    this::accumulateLocally
```

Comment:
Why do we have to be so tricky here? Why not leave the hit context preparation as is and simply do the accounting below after the fetch processors have done their work? That gives us a much cleaner accounting in case of filtering as well while making this change a lot simpler?

Reply:
We need to hook in the accounting math in the cases where we lazily load the source https://github.com/elastic/elasticsearch/pull/121920/files#diff-e05d7cea3ce24b7a468186919b40a9647ff7c3390a4434a2e36e3302d77ca4bbR397

Reply:
Yes but after the fetch processors ran we don't even escape the `HitContext`, we're definitely done with the source there one way or another?

Reply:
How would you set the memory ref counted thing on the hit _after_ it was created in this case?